### PR TITLE
Introduce state mutability (to replace const/payable)

### PR DESCRIPTION
**Pull request by axic**
Depends on #2624. Part of #992.

### CODE REVIEW

File: libsolidity/ast/AST.h [hunk 1]
```diff
@@ -152,6 +152,9 @@ class Declaration: public ASTNode
 	/// Visibility ordered from restricted to unrestricted.
 	enum class Visibility { Default, Private, Internal, Public, External };
 
+	// How the function can mutate the EVM state.
+	enum class StateMutability { Pure, View, NonPayable, Payable };
```
Comment (axic) [inline libsolidity/ast/AST.h]: Placing it here will not work due to forward declarations hiding it. Any suggestions for where to place this?

File: libsolidity/ast/AST.h [hunk 2]
```diff
@@ -566,32 +569,31 @@ class FunctionDefinition: public CallableDeclaration, public Documented, public
...
 	virtual void accept(ASTVisitor& _visitor) override;
 	virtual void accept(ASTConstVisitor& _visitor) const override;
 
+	StateMutability stateMutability() const { return m_stateMutability; }
 	bool isConstructor() const { return m_isConstructor; }
 	bool isFallback() const { return name().empty(); }
-	bool isDeclaredConst() const { return m_isDeclaredConst; }
-	bool isPayable() const { return m_isPayable; }
+	bool isDeclaredConst() const { return m_stateMutability == StateMutability::View; }
+	bool isPayable() const { return m_stateMutability == StateMutability::Payable; }
```
Comment (axic) [inline libsolidity/ast/AST.h:612]: These can be replaced too at a later stage, once the basics are done properly.

**Comment by axic** [comment]
@chriseth I think this is close to a reviewable state now. Monster change but after it readability increased in a lot of places.

File: libsolidity/ast/Types.cpp [hunk 3]
```diff
@@ -2292,9 +2287,9 @@ string FunctionType::toString(bool _short) const
 	for (auto it = m_parameterTypes.begin(); it != m_parameterTypes.end(); ++it)
 		name += (*it)->toString(_short) + (it + 1 == m_parameterTypes.end() ? "" : ",");
 	name += ")";
-	if (m_isConstant)
+	if (isConstant())
 		name += " constant";
```
Comment (axic) [inline libsolidity/ast/Types.cpp:2306]: What is this `toString` function used for? If we change the parser to support `view` next to `constant`, but only print `view`, would that had any adverse effect? I think it is only used in error reporting.

File: libsolidity/analysis/TypeChecker.cpp [hunk 4]
```diff
@@ -84,7 +84,7 @@ bool TypeChecker::visit(ContractDefinition const& _contract)
 	{
 		if (!function->returnParameters().empty())
 			m_errorReporter.typeError(function->returnParameterList()->location(), "Non-empty \"returns\" directive for constructor.");
-		if (function->isDeclaredConst())
+		if (function->stateMutability() == StateMutability::View)
```
Comment (axic) [inline libsolidity/analysis/TypeChecker.cpp]: So should we have this or a `isView()` helper?

**Comment by axic** [comment]
@chriseth where to place a helper to print the text version of state mutability? It will be used in Parser, Types::identifier, ABI.

File: libsolidity/ast/Types.cpp [hunk 5]
```diff
@@ -2000,8 +2000,7 @@ TypePointer TupleType::closestTemporaryType(TypePointer const& _targetType) cons
 
 FunctionType::FunctionType(FunctionDefinition const& _function, bool _isInternal):
 	m_kind(_isInternal ? Kind::Internal : Kind::External),
-	m_isConstant(_function.isDeclaredConst()),
-	m_isPayable(_isInternal ? false : _function.isPayable()),
+	m_stateMutability(_function.stateMutability()),
```
Comment (axic) [inline libsolidity/ast/Types.cpp:2003]: Do you remember the reason for the extra `_isInternal ? false : _function.isPayable()` check here?

Comment (chriseth) [reply libsolidity/ast/Types.cpp:2003]: This constructor is used to create either the internal or the external view of a public function. If the internal view is requested, then the function type is of course not payable.

Comment (axic) [reply libsolidity/ast/Types.cpp:2003]: So using here `_isInternal ? StateMutability::NonPayable : _function.stateMutability()`.

File: libsolidity/analysis/TypeChecker.cpp (hunk 4 above)
Comment (axic) [reply libsolidity/analysis/TypeChecker.cpp]: I keep `isDeclaredConst()` for the moment.

File: libsolidity/ast/ASTForward.h [hunk 6]
```diff
@@ -95,6 +95,8 @@ using ASTPointer = std::shared_ptr<T>;
 
 using ASTString = std::string;
 
+// How the function can mutate the EVM state.
+enum class StateMutability { Pure, View, NonPayable, Payable };
```
Comment (axic) [inline libsolidity/ast/ASTForward.h]: Let's create an `ASTENums.h` and move this  `stateMutabilityToString()` or overload `to_string()`.

File: libsolidity/ast/Types.cpp [hunk 7]
```diff
@@ -2000,8 +2000,7 @@ TypePointer TupleType::closestTemporaryType(TypePointer const& _targetType) cons
 
 FunctionType::FunctionType(FunctionDefinition const& _function, bool _isInternal):
 	m_kind(_isInternal ? Kind::Internal : Kind::External),
-	m_isConstant(_function.isDeclaredConst()),
-	m_isPayable(_isInternal ? false : _function.isPayable()),
+	m_stateMutability(_isInternal ? StateMutability::NonPayable : _function.stateMutability()),
```
Comment (axic) [inline libsolidity/ast/Types.cpp]: This will need to be more precise because it changes view to non payable.

**Comment by axic** [comment]
@chriseth should be ready

File: libsolidity/ast/ASTEnums.h [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+namespace dev
+{
+namespace solidity
+{
+
+// How a function can mutate the EVM state.
+enum class StateMutability { View, NonPayable, Payable };
+
+inline std::string stateMutabilityToString(StateMutability const& _stateMutability)
```
Comment (chriseth) [inline libsolidity/ast/ASTEnums.h:36]: could be just named `toString` or `to_string` - this is C++. Also the argument can be passed by value.

File: libsolidity/ast/Types.cpp (hunk 5 above)
Comment (chriseth) [reply libsolidity/ast/Types.cpp:2003]: no ,if it was `view` it should stay view.

`_isInternal && _function.stateMutability() == StateMutability::Payable ? StateMutability::NonPayable : _function.stateMutability()`.

Comment (chriseth) [reply libsolidity/ast/Types.cpp:2003]: Ah you have it below :-) 👍

File: libsolidity/ast/Types.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/ast/Types.cpp:2306]: I think so, too.

**Review by chriseth — APPROVED** [review]

File: libsolidity/ast/ASTEnums.h (hunk 8 above)
Comment (axic) [reply libsolidity/ast/ASTEnums.h:36]: For some reason using `to_string` and `toString` clashes with std that callers of `to_string(1)`.

Comment (axic) [inline libsolidity/ast/ASTEnums.h]: Okay can use `toString` as long as I call using the namspace `solidity::toString()`

Comment (axic) [reply libsolidity/ast/ASTEnums.h:36]: Let us fix this in a separate PR.

**Pull request merged**